sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/BoundAttribute.scala

```diff
@@ -65,6 +65,40 @@ case class BoundReference(ordinal: Int, dataType: DataType, nullable: Boolean)
   }
 }
 
+/**
+ * This bound reference points to a parameterized slot in an input tuple. It is used in
+ * common sub-expression elimination. When some common sub-expressions have same structural
+ * but different slots of input tuple, we replace `BoundReference` with this parameterized
+ * version. The slot position is parameterized and is given at runtime.
+ */
+case class ParameterizedBoundReference(parameter: String, dataType: DataType, nullable: Boolean)
+  extends LeafExpression {
+
+  override def toString: String = s"input[$parameter, ${dataType.simpleString}, $nullable]"
+
+  override def eval(input: InternalRow): Any = {
+    throw new UnsupportedOperationException(
+      "ParameterizedBoundReference does not implement eval")
+  }
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    assert(ctx.currentVars == null && ctx.INPUT_ROW != null,
+      "ParameterizedBoundReference can not be used in whole-stage codegen yet.")
```

Comment:
Any barrier to support the whole-stage codegen case?

Reply:
I think it can be used in whole-stage codegen. As this is applied to sub-expression elimination which is non whole-stage only, I'd also like to reduce the change of code in a single PR.